### PR TITLE
feat: add columns resizable options

### LINKED ISSUE #2936: DataTable: Add resizable columns
Implement https://www.lightningdesignsystem.com/components/data-tables/#Column-resizing 

Functionality should match: https://developer.salesforce.com/docs/component-library/bundle/lightning-datatable/documentation

> ### Toggle Between Navigation and Action Modes
> You can use data tables in navigation mode and action mode using the keyboard. To enter navigation mode, tab into the data table, which triggers focus on the first data cell in the table body. Use the arrow keys to move around the table.
To enter action mode, press the Enter key or Space Bar. Columns can be resized in action mode. To resize a column, navigate to the header by pressing the Up Arrow key. Then, press the Tab key to activate the column divider, and resize the column using the Left Arrow and Right Arrow keys. To finish resizing the column and return to navigation mode, press the Tab key. 

Main callouts not in SLDS spec are:
* Column handler highlights when focused _and_ hover
* Focused column handler keyboard events on header cell when in action mode: left and right arrow keys should resize column +10px, -10px

For Design System React, this means that the column tab/resize keyboard events only happen when `keyboardNavigation:true` . For an example of this action mode in DSR, hit enter on a cell here: http://design-system-react-components.herokuapp.com/?path=/story/sldsdatatable--interactive-elements and an example of the completed implementation here: https://developer.salesforce.com/docs/component-library/bundle/lightning-datatable/example
When the header cell is “selected” in action mode (enter), then the header resize column button gets a tab stop and the left and right arrows get actions. When escape is pressed, keyboard events and tabIndexes go away, but the mouse can still resize. `keyboardNavigation:true` and action mode are already implemented. We only need a `tabindex` changed to `0` when in "action mode" and a keyboard arrow event that resizes on left and right added when that tabstop is focused.

### Video of keyboard interaction
Please use the instruction above to reproduce

### Screenshot of mouse _and_ keyboard event highlight
Screen Shot 2021-07-21 at 12 11 47 PM

### `tabIndex` change when action mode is toggled with enter key:

Screen Shot 2021-07-22 at 11 41 36 AM
Screen Shot 2021-07-22 at 11 41 15 AM

### PR DESCRIPTION
**Pull request by mmarchett**
Resolves: B2BSOLUT-5767

Fixes #2936

### Additional description
Added the functionality to modify the width of the columns in the data-table component.
---

### CONTRIBUTOR checklist (do not remove)
Please complete for every pull request

* [x] First-time contributors should sign the Contributor License Agreement. It's a fancy way of saying that you are giving away your contribution to this project. If you haven't before, wait a few minutes and a bot will comment on this pull request with instructions.
* [X] `npm run lint:fix` has been run and linting passes.
* [X] Mocha, Jest (Storyshots), and `components/component-docs.json` CI tests pass (`npm test`).
* [X] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [X] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [X] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [X] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).

### REVIEWER checklist (do not remove)

* [ ] CircleCI tests pass. This includes linting, Mocha, Jest, Storyshots, and `components/component-docs.json` tests.
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [ ] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).
###### Required only if there are markup / UX changes
* [ ] Add year-first date and commit SHA to `last-slds-markup-review` in `package.json` and push.
* [ ] Request a review of the deployed Heroku app by the Salesforce UX Accessibility Team.
* [ ] Add year-first review date, and commit SHA, `last-accessibility-review`, to `package.json` and push.
* [ ] While the contributor's branch is checked out, run `npm run local-update` within locally cloned [site repo](https://github.com/salesforce-ux/design-system-react-site) to confirm the site will function correctly at the next release.

### CODE REVIEW

**Comment by interactivellama** [comment]
I've started an issue and spec for this feature. 
Please add the following two bullet points to the PR https://github.com/salesforce/design-system-react/issues/2936

**Comment by interactivellama** [comment]
All the column cells have a tab stop now which is great, but that should only be the case in action/edit mode. The DataTable should only have one tab stop for the whole table when in Navigation mode. Now all cells have a tab index when in Navigation mode.

The tab events within a cell are broken now. Tab key now takes the user to the next column even if there are multiple tabbable items within a cell. It should only tab to the next column after tabbing through all tabbable items in a cell. See Opportunity Name column on http://localhost:9001/?path=/story/sldsdatatable--interactive-elements when in edit mode.

**Review by interactivellama — CHANGES_REQUESTED** [review]
I did a quick first pass of the code and left a few comments. This is looking great! We are getting really close. 

Due to the "double header" way that the fixed layout feature is implemented, we will want to set the width of the "second header table column" to the same width on initial resize. It works when the table is scrolled due to events which is great, but isn't sync'd initially when columns are resized when `fixedLayout/fixedHeader` is enabled. See below:

Screen Shot 2021-07-29 at 7 39 10 AM

Does your usage/UX need non-scrollable tables to be column resizable? After reviewing the examples, I believe it would be best to limit resizing to only `fixedLayout/fixedHeader`. This isn't something I realized early on. Does that work for your use case? @mmarchett  This PR will make the usage more predictable and limit what we need to support/debug in the future! https://github.com/mmarchett/design-system-react/pull/2

For instance, https://developer.salesforce.com/docs/component-library/bundle/lightning-datatable/example doesn't even allow a non-scrollable/basic table.

File: components/data-table/index.jsx [hunk 1]
```diff
@@ -356,10 +366,34 @@ class DataTable extends React.Component {
...
 	componentWillUnmount() {
 		this.toggleFixedHeaderListeners(false);
+
+		if (this.props.resizable) {
+			this.disableResize();
+		}
+	}
+
+	onResize() {
+		const table = document.querySelector(`#${this.getId()}`);
```
[the remark this answers is not in the document]
Comment (mmarchett) [reply components/data-table/index.jsx]: All references to document have been replaced by refs.

File: components/data-table/index.jsx [hunk 2]
```diff
@@ -356,10 +366,34 @@ class DataTable extends React.Component {
...
+		const columns = this.props.fixedHeader
+			? table.getElementsByClassName('slds-cell-fixed')
```
[the remark this answers is not in the document]
Comment (mmarchett) [reply components/data-table/index.jsx]: All references to document have been replaced by refs

File: components/data-table/index.jsx [hunk 3]
```diff
@@ -500,10 +545,68 @@ class DataTable extends React.Component {
...
+				this.gripRefs.forEach((grip) => {
+					// eslint-disable-next-line no-param-reassign
+					grip.style.height = `${parseInt(grip.style.height, 10) + 33}px`;
+				});
+			}
+		}
+	}
+
+	enableResize() {
+		const remoteTable = document.querySelector(`#${this.getId()}`);
```
[the remark this answers is not in the document]
Comment (mmarchett) [reply components/data-table/index.jsx]: All references to document have been replaced by refs

File: components/data-table/private/header-cell.jsx [hunk 4]
```diff
@@ -233,6 +240,7 @@ class DataTableHeaderCell extends React.Component {
 							);
 							return (
 								<th
+									id={property}
```
[the remark this answers is not in the document]
Comment (mmarchett) [reply components/data-table/private/header-cell.jsx]: The generation of the id was modified to be unique.

File: components/data-table/index.jsx [hunk 5]
```diff
@@ -66,6 +67,11 @@ const defaultProps = {
...
 	hasMore: false,
 	loadMoreOffset: 20,
+	resizable: false,
+	resizerOptions: {
+		resizeMode: 'flex',
+		draggingClass: 'slds-table-column-resizer',
+	},
 };
 
 /**
```
[the remark this answers is not in the document]
Comment (mmarchett) [reply components/data-table/index.jsx:77]: Added documentation for resizable property and resizerOptions.

**Comment by mmarchett** [comment]
> I did a quick first pass of the code and left a few comments. This is looking great! We are getting really close.
> 
> Due to the "double header" way that the fixed layout feature is implemented, we will want to set the width of the "second header table column" to the same width on initial resize. It works when the table is scrolled due to events which is great, but isn't sync'd initially when columns are resized when `fixedLayout/fixedHeader` is enabled. See below:
> 
> Screen Shot 2021-07-29 at 7 39 10 AM
> 
> Does your usage/UX need non-scrollable tables to be column resizable? After reviewing the examples, I believe it would be best to limit resizing to only `fixedLayout/fixedHeader`. This isn't something I realized early on. Does that work for your use case? @mmarchett This PR will make the usage more predictable and limit what we need to support/debug in the future! [mmarchett#2](https://github.com/mmarchett/design-system-react/pull/2)
> 
> For instance, https://developer.salesforce.com/docs/component-library/bundle/lightning-datatable/example doesn't even allow a non-scrollable/basic table.

Thanks for the feedback, as you advised, now the resize functionality would be available in conjunction with fixedLayout only, as specified at https://react.lightningdesignsystem.com/components/data-tables/

File: components/data-table/__tests__/data-table.browser-test.jsx [hunk 6]
```diff
@@ -704,6 +704,47 @@ describe('DataTable: ', function describeFunction() {
...
+						onAction={() => {}}
+						dropdown={<Dropdown length="5" />}
+					/>
+				</DataTable>
+			).call(this);
+
+			expect(this.dom.querySelectorAll('.grip-resizable').length).to.eql(5);
+			expect(this.dom.querySelectorAll('.grip-container').length).to.eql(1);
+		});
 	});
```
Comment (interactivellama) [inline components/data-table/__tests__/data-table.browser-test.jsx:748]: We need at least two tests that will show that resize works. One left key pressed and one right key. Then we should be good to go.

**Review by interactivellama — CHANGES_REQUESTED** [review]
This is looking great! UX is good. I left one request for tests.

File: components/data-table/__tests__/data-table.browser-test.jsx (hunk 6 above)
Comment (mmarchett) [reply components/data-table/__tests__/data-table.browser-test.jsx:748]: I just added tests for these scenarios!

**Review by interactivellama — APPROVED** [review]
Looks good!

**Pull request merged**